### ISSUE TITLE
ca65: Endless loop printing error messages

### ISSUE THREAD

**Issue by kugelfuhr** [cc65/cc65#1678]
The following input causes the assembler to go into an endless loop printing error messages:

```
        .macro  foo
                foo
        .endmacro
        .struct bar
                foo
        .endstruct
```

**Comment by spiro-trikaliotis** [comment]
Interesting:

```
macro-loop-struct.s:5: Note: Dropping 248 additional line infos
macro-loop-struct.s:5: Error: Too many nested macro expansions
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:5: Note: Dropping 248 additional line infos
macro-loop-struct.s:5: Error: Too many nested macro expansions
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here
macro-loop-struct.s:2: Note: Macro was defined here

```

ca65 recognizes that there are too many nesting levels, but it cannot quit the loop.

**Comment by kugelfuhr** [comment]
This patch might work:

```diff
diff --git a/src/ca65/macro.c b/src/ca65/macro.c
index d6c807035..1dee4c0af 100644
--- a/src/ca65/macro.c
+++ b/src/ca65/macro.c
@@ -801,9 +801,6 @@ static void StartExpClassic (MacExp* E)
 {
     token_t Term;
 
-    /* Skip the macro name */
-    NextTok ();
-
     /* Does this invocation have any arguments? */
     if (!TokIsSep (CurTok.Tok)) {
 
@@ -887,9 +884,6 @@ static void StartExpDefine (MacExp* E)
     */
     unsigned Count = E->M->ParamCount;
 
-    /* Skip the current token */
-    NextTok ();
-
     /* Read the actual parameters */
     while (Count--) {
         TokNode* Last;
@@ -964,15 +958,20 @@ static void StartExpDefine (MacExp* E)
 
 void MacExpandStart (Macro* M)
 /* Start expanding a macro */
-{
+{   
+    FilePos Pos;
     MacExp* E;
 
     /* Check the argument */
     PRECONDITION (M && (M->Style != MAC_STYLE_DEFINE || DisableDefines == 0));
 
+    /* Remember the current file position, then skip the macro name token */
+    Pos = CurTok.Pos;
+    NextTok ();
+
     /* We cannot expand an incomplete macro */
     if (M->Incomplete) {
-        Error ("Cannot expand an incomplete macro");
+        PError (&Pos, "Cannot expand an incomplete macro");
         return;
     }
 
@@ -980,7 +979,7 @@ void MacExpandStart (Macro* M)
     ** to force an endless loop and assembler crash.
     */
     if (MacExpansions >= MAX_MACEXPANSIONS) {
-        Error ("Too many nested macro expansions");
+        PError (&Pos, "Too many nested macro expansions");
         return;
     }
```

**Comment by spiro-trikaliotis** [comment]
Well, I have thought about something similar, yes: Advancing the token in the case of an error.

However, as I do not fully understand that code, and we do not have an appropriate testsuite here, I am reluctant in changing such behaviour.

Is here someone else with an opinion? - other than adding a proper test suite? ;)

**Comment by mrdudz** [comment]
> adding a proper test suite

this is the way to go :)

**Comment by spiro-trikaliotis** [comment]
> > adding a proper test suite
> 
> this is the way to go :)

That's why I started PR #1679.

**Comment by kugelfuhr** [comment]
> However, as I do not fully understand that code, and we do not have an appropriate testsuite here, I am reluctant in changing such behaviour.

So let me analyze the bug for you. After that, I will have a look at the supplied patch.

**What is the problem with the input?**
The input contains a recursive macro that has no provisions for terminating recursion. The assembler should and does spot that and terminates macro expansion with an error message.

**Where does the endless loop come from?**
The struct parser checks for each struct member if it is a macro name. If so it calls a function to expand this macro. But if this function encounters an error (as in the given case) it will not skip the macro name. So once back in the struct parser, the assembler will read the same macro name again. And will try to expand it. And again. And so on. This causes an endless loop, and in every loop iteration the assembler will complain once that it cannot expand the macro.

**So what is the fix?**
From the paragraph above, the fix is obvious. The macro name has to be skipped in case of an error so a function that calls the macro expansion can continue with the tokens after the macro, regardless if expansion was successful or not.

**Where does this have to happen?**
All functions that call the macro expansion code expect the macro to be expanded and the macro tokens to be skipped so
they can process the following tokens. So obviously this routine should also skip the macro name in case of an error. Callers then don't have to care if expansion was successful or not, they can simply proceed with processing the following tokens.

With this analysis in mind, let's look at the patch:

- The patch changes the function `MacExpandStart()` which is the function called to expand a macro. This is in line with the analysis above.
- `MacExpandStart()` calls one of two other functions (`StartExpClassic()` or `StartExpDefine()`) depending on the type of the macro. Both of these functions will skip the macro name as their first action.
- The patch removes the calls to `NextTok()` in these two functions and pulls it into `MacExpandStart()` instead. So for the "no error" case nothing has actually changed. The macro name token is skipped before further processing occurs.
- But in the "error" case, the behavior differs now, because the macro name is skipped before checking for errors. And this is actually what is necessary to fix the bug as shown in the analysis above.
  
That wasn't too difficult, was it?

**Comment by mrdudz** [comment]
That last post was convincing enough for me - applied the fix in b6ffa4af64914ed9b5805321a68576e3a785931b

Please test

**Comment by mrdudz** [comment]
no feedback or complaints -> closing
